src/vos/vos_pool_scrub.c

```diff
@@ -0,0 +1,735 @@
+				*acts |= VOS_ITER_CB_YIELD;
+				ctx->sc_did_yield = false;
+			}
+
+			if (rc != 0) {
+				D_ERROR("Error Verifying:"DF_RC"\n", DP_RC(rc));
+				return rc;
+			}
+			ds_scrub_sched_control(ctx);
+		}
+		break;
+	}
+
+	case VOS_ITER_DTX:
+	case VOS_ITER_COUUID:
+	case VOS_ITER_NONE:
+		D_ASSERTF(0, "Invalid.");
+	}
+
+	return 0;
+}
+
+static int
+sc_scrub_cont(struct scrub_ctx *ctx)
+{
+	vos_iter_param_t		param = {0};
+	struct vos_iter_anchors		anchor = {0};
+	int				rc;
+
+	/* not all containers in the pool will have checksums enabled */
+	if (!daos_csummer_initialized(sc_csummer(ctx)))
+		return 0;
+
+	C_TRACE("Scrubbing container '"DF_UUIDF"'\n",
+		DP_UUID(*sc_cont_uuid(ctx)));
+
+	param.ip_hdl = sc_cont_hdl(ctx);
+	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+	rc = vos_iterate(&param, VOS_ITER_OBJ, true, &anchor,
+			 obj_iter_scrub_pre_cb,
+			 NULL, ctx, NULL);
+
+	if (rc != DER_SUCCESS) {
+		D_ERROR("Object scrub failed: "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	ds_scrub_sched_control(ctx);
+
+	return 0;
+}
+
+static int
+sc_cont_setup(struct scrub_ctx *ctx, vos_iter_param_t *param,
+	      vos_iter_entry_t *entry)
+{
+	int			 rc;
+
+	if (ctx->sc_cont_lookup_fn == NULL)
+		return -DER_NOSYS;
+	rc = ctx->sc_cont_lookup_fn(ctx->sc_pool_uuid, entry->ie_couuid,
+				    ctx->sc_sched_arg, &ctx->sc_cont);
+
+	if (rc != 0) {
+		D_ERROR("Error opening vos container: "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	return 0;
+}
+
+static void
+sc_cont_teardown(struct scrub_ctx *ctx)
+{
+	if (ctx->sc_cont_put_fn != NULL)
+		ctx->sc_cont_put_fn(ctx->sc_cont.scs_cont_src);
+	memset(&ctx->sc_cont, 0, sizeof(ctx->sc_cont));
+}
+
+/** vos_iter_cb_t */
+static int
+cont_iter_scrub_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		   vos_iter_type_t type, vos_iter_param_t *param,
+		   void *cb_arg, unsigned int *acts)
+{
+	struct scrub_ctx	*ctx = cb_arg;
+	int			 rc = 0;
+
+	D_ASSERT(type == VOS_ITER_COUUID);
+
+	rc = sc_cont_setup(ctx, param, entry);
+	if (rc != 0)
+		return rc;
+
+	D_DEBUG(DB_CSUM, "Scrubbing container: "DF_UUID"\n",
+		DP_UUID(ctx->sc_cont.scs_cont_uuid));
+
+	rc = sc_scrub_cont(ctx);
+
+	sc_cont_teardown(ctx);
+
+	return rc;
+}
+
+static void
+sc_pool_start(struct scrub_ctx *ctx)
+{
+	/* remember previous checksum calculations */
+	ctx->sc_pool_last_csum_calcs = ctx->sc_pool_csum_calcs;
+	ctx->sc_pool_csum_calcs = 0;
+	d_gettime(&ctx->sc_pool_start_scrub);
+	ctx->sc_status = SCRUB_STATUS_RUNNING;
+
+	sc_m_pool_csum_reset(ctx);
+	sc_m_pool_start(ctx);
+}
+
+int
+ds_scrub_pool(struct scrub_ctx *ctx)
+{
+	vos_iter_param_t	param = {0};
+	struct vos_iter_anchors	anchor = {0};
+	int			rc;
+
+	if (daos_handle_is_inval(ctx->sc_vos_pool_hdl)) {
+		D_ERROR("vos_iter_handle is invalid.\n");
+		return -DER_INVAL;
+	}
+
+	sc_pool_start(ctx);
+
+	param.ip_hdl = ctx->sc_vos_pool_hdl;
+	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+	rc = vos_iterate(&param, VOS_ITER_COUUID, false, &anchor,
+			 NULL, cont_iter_scrub_cb, ctx, NULL);
```

Comment:
The problem of using two vos_iterate() calls (one for container level, the other for object and below levels) is that cont_iter_scrub_cb() has no idea if it yielded in object iterating, so if container tree is changed by others (created container or destroyed container) in between, container level iterating may run into trouble without re-probe.

You'd blindly set YIELD flag in cont_iter_scrub_cb() to avoid the potential problem.

Reply:
Am blindly setting YIELD flag in cont_iter_scrub_cb(). Might try to make smarter later.